diesel/src/connection/statement_cache.rs

```diff
@@ -111,3 +140,23 @@ impl<DB> StatementCacheKey<DB> where
         Ok(query_builder.finish())
     }
 }
+
+fn refmut_map_result<T, U, F>(refmut: RefMut<T>, f: F)
+    -> QueryResult<RefMut<U>> where
+        F: FnOnce(&mut T) -> QueryResult<&mut U>,
+{
+    use std::mem;
+
+    let mut error = None;
+    let result = RefMut::map(refmut, |mutref| match f(mutref) {
+        Ok(x) => x,
+        Err(e) => {
+            error = Some(e);
+            unsafe { mem::uninitialized() }
+        }
+    });
+    match error {
+        Some(e) => Err(e),
+        None => Ok(result),
+    }
+}
```

Comment:
I know we talked about this at length but I' still looking for a version that is just one bit nicer. Good for now, though. (I think `std` should have more `try_` methods. `try_into` is a thing, and `try_map`, `try_unwrap_or_else`, et.al. should be too)